staging/src/k8s.io/apiserver/pkg/cel/mutation/typeprovider.go

```diff
@@ -0,0 +1,92 @@
+/*
+Copyright 2024 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package mutation
```

Comment:
I think this is an OK package for this, at least for now.   Long term, it might be reasonable to allow validation features to leverage object initialization. for example `spec.something == Object.spec.something{a: 1, b: true}` could be convenient.

Reply:
I believe implementation of `DeepEqual` would involve the second-stage type resolver. The `==` operator requires `lhs` and `rhs` to be of the same type. source: https://github.com/google/cel-spec/blob/master/doc/langdef.md#equality

To know the type inside the existing object referred as `spec.something`, we will have to look up the schema. Or assume `spec.something` is of `Object.something` type, which may or may not be true.

Reply:
The spec will permit a `dyn == T` comparison, and at runtime a `DeepEqual` check could ensure that the types agree, and are equal. If the values are not of the same type at type-check time, this will surface an error at config time.

Reply:
Good point. Added support for equality check